### PR TITLE
Add support for `inline Cpp` declarations.

### PR DESCRIPTION
**Pull request by zygoloid**
For #6830, add support for inline C++ fragments as a declaration rather than as a packaging directive. For now, this uses `inline Cpp <string-literal>;` as syntax. The prior `import Cpp inline <string-literal>;` is left alone for the time being. We can decide separately whether to remove that.

`inline Cpp` requires that there was at least one `import Cpp`. It's not clear to me if that's the right design long-term, but it seems reasonable for now.

Assisted-by: Gemini via Google Antigravity

### CODE REVIEW

File: toolchain/sem_ir/inst_kind.def [hunk 1]
```diff
@@ -87,6 +87,7 @@ CARBON_SEM_IR_INST_KIND(ImplWitnessAccessSubstituted)
 CARBON_SEM_IR_INST_KIND(ImplWitnessAssociatedConstant)
 CARBON_SEM_IR_INST_KIND(ImplWitnessTable)
 CARBON_SEM_IR_INST_KIND(ImportCppDecl)
+CARBON_SEM_IR_INST_KIND(InlineCppDecl)
```
Comment (danakj) [inline toolchain/sem_ir/inst_kind.def]: Sort?

File: toolchain/parse/typed_nodes.h [hunk 2]
```diff
@@ -251,6 +251,18 @@ struct InlineImportSpecifier {
   InlineImportBodyId body;
 };
```
Comment (danakj) [inline toolchain/parse/typed_nodes.h:252]: Do we still support this too?

File: toolchain/parse/testdata/packages/import/cpp_inline.carbon [hunk 3]
```diff
@@ -150,3 +198,50 @@ int n;
...
+// CHECK:STDOUT:     {kind: 'FileStart', text: ''},
+// CHECK:STDOUT:       {kind: 'InlineIntroducer', text: 'inline'},
+// CHECK:STDOUT:     {kind: 'InlineCppDecl', text: ';', has_error: yes, subtree_size: 2},
+// CHECK:STDOUT:       {kind: 'InlineIntroducer', text: 'inline'},
+// CHECK:STDOUT:     {kind: 'InlineCppDecl', text: ';', has_error: yes, subtree_size: 2},
+// CHECK:STDOUT:       {kind: 'InlineIntroducer', text: 'inline'},
+// CHECK:STDOUT:       {kind: 'CppNameExpr', text: 'Cpp'},
+// CHECK:STDOUT:     {kind: 'InlineCppDecl', text: ';', has_error: yes, subtree_size: 3},
+// CHECK:STDOUT:       {kind: 'InlineIntroducer', text: 'inline'},
+// CHECK:STDOUT:       {kind: 'CppNameExpr', text: 'Cpp'},
```
Comment (danakj) [inline toolchain/parse/testdata/packages/import/cpp_inline.carbon:221]: Is the subtree_size wrong here?

File: toolchain/parse/testdata/packages/import/cpp_inline.carbon [hunk 4]
```diff
@@ -72,6 +72,54 @@ import Cpp inline '''
...
+// CHECK:STDERR:
+// CHECK:STDERR: ^
+// CHECK:STDERR:
+
+// --- todo_fail_nested.carbon
+
+class C {
+  inline Cpp "";
+}
+
```
Comment (danakj) [inline toolchain/parse/testdata/packages/import/cpp_inline.carbon]: Previously we said we want parse to accept things from the wrong context and we should fail in check. We could make parse reject this but we only use parse differentiating context when we need to produce a different node based on context (like for let). So IIUC, this test should not be a fail. It's not clear if we need it or not, but we it should be a todo_fail in check. Same for putting it in an interface, or fn.

File: toolchain/check/testdata/interop/cpp/inline_decl.carbon [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/none.carbon
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/interop/cpp/inline_decl.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/interop/cpp/inline_decl.carbon
+
+// --- inline_decl.carbon
```
Comment (danakj) [inline toolchain/check/testdata/interop/cpp/inline_decl.carbon:13]: Could we test two inline blocks that both define the same function name? That should produce an error iiuc.

File: toolchain/check/handle_inline_decl.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    return true;
+  }
+  CARBON_CHECK(context.cpp_context(), "Have `Cpp` name but no Cpp context");
+
+  auto string_token = context.parse_tree().node_token(body_id);
+  auto string_value_id = context.tokens().GetStringLiteralValue(string_token);
+  auto string_value = context.string_literal_values().Get(string_value_id);
+
+  InjectAstFromInlineCode(context, SemIR::LocId(node_id), string_value);
+  AddInst<SemIR::InlineCppDecl>(context, node_id, {.text_id = string_value_id});
```
Comment (danakj) [inline toolchain/check/handle_inline_decl.cpp]: Should the location be the body, excluding the Cpp keyword?

File: toolchain/check/cpp/generate_ast.cpp [hunk 7]
```diff
@@ -686,6 +708,43 @@ auto GenerateAst(Context& context,
   return true;
 }
 
+auto InjectAstFromInlineCode(Context& context, SemIR::LocId loc_id,
+                             llvm::StringRef source_code) -> bool {
+  auto* cpp_context = context.cpp_context();
+  if (!cpp_context) {
+    return false;
+  }
```
Comment (danakj) [inline toolchain/check/cpp/generate_ast.cpp]: I think we CHECKed this at the caller?

File: toolchain/check/handle_inline_decl.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      SemIR::ErrorInst::ConstantId) {
+    return true;
+  }
+  CARBON_CHECK(context.cpp_context(), "Have `Cpp` name but no Cpp context");
+
+  auto string_token = context.parse_tree().node_token(body_id);
+  auto string_value_id = context.tokens().GetStringLiteralValue(string_token);
+  auto string_value = context.string_literal_values().Get(string_value_id);
+
+  InjectAstFromInlineCode(context, SemIR::LocId(node_id), string_value);
```
Comment (danakj) [inline toolchain/check/handle_inline_decl.cpp]: Should we be handling `false` return here? Would it result in an ErrorInst anywhere?

File: toolchain/sem_ir/inst_kind.def (hunk 1 above)
Comment (zygoloid) [reply toolchain/sem_ir/inst_kind.def]: Done.

File: toolchain/parse/testdata/packages/import/cpp_inline.carbon (hunk 3 above)
Comment (zygoloid) [reply toolchain/parse/testdata/packages/import/cpp_inline.carbon:221]: The tree is printed in postorder -- the children are above the parent.

File: toolchain/parse/typed_nodes.h (hunk 2 above)
Comment (zygoloid) [reply toolchain/parse/typed_nodes.h:252]: Yes, for now at least. (I called this out in the PR description.) I'm not sure if we'll want to remove it or not -- it's the only way to inject code between imported headers without factoring it out to a separate file, and keeping it avoids breaking any of our existing tests. Something to iterate on, I think.

File: toolchain/parse/testdata/packages/import/cpp_inline.carbon (hunk 4 above)
Comment (zygoloid) [reply toolchain/parse/testdata/packages/import/cpp_inline.carbon]: Yeah. I mostly added this test as a reminder to myself that we should be doing *something* to reject this case, but I agree that we probably want to reject it in check. And you know what? Adding a todo test in check seems about the same amount of effort as actually implementing a diagnostic for this, so I'll just do that :)

File: toolchain/check/handle_inline_decl.cpp (hunk 6 above)
Comment (zygoloid) [reply toolchain/check/handle_inline_decl.cpp]: It should. Fixed, and added a test that we get the diagnostic location within the inline C++ code right.

File: toolchain/check/testdata/interop/cpp/inline_decl.carbon (hunk 5 above)
Comment (zygoloid) [reply toolchain/check/testdata/interop/cpp/inline_decl.carbon:13]: Done.

File: toolchain/check/cpp/generate_ast.cpp (hunk 7 above)
Comment (zygoloid) [reply toolchain/check/cpp/generate_ast.cpp]: Replaced with a CHECK.

File: toolchain/check/handle_inline_decl.cpp (hunk 8 above)
Comment (zygoloid) [reply toolchain/check/handle_inline_decl.cpp]: The only reason this returns false is if clang's `EnterSourceFile` fails, and that only happens if it can't find a memory buffer for the file, which should be impossible since we provided it with a memory buffer. So it's not actually possible for this to fail currently. (But who knows how Clang's implementation will change over time.)

If this returns false, a diagnostic will have been produced (by Clang rather than directly by us), and I wouldn't expect there to be an `ErrorInst` anywhere. I don't know, maybe this should have a `void` return type since there's not really anything the caller can do about it failing, and we already diagnosed it in the callee. Do you have a preference?

Comment (danakj) [reply toolchain/check/handle_inline_decl.cpp]: Yeah if the caller can't do anything, I would prefer `void`. We can be clear that we're discarding failure in the body of `InjectAstFromInlineCode`

File: toolchain/parse/testdata/packages/import/cpp_inline.carbon (hunk 3 above)
Comment (danakj) [reply toolchain/parse/testdata/packages/import/cpp_inline.carbon:221]: Oh the newlines in the text block really confused me oops

File: toolchain/parse/typed_nodes.h (hunk 2 above)
Comment (danakj) [reply toolchain/parse/typed_nodes.h:252]: Oh woops, sorry for not reading the description.

**Review by danakj — APPROVED** [review]
LGTM % void return

File: toolchain/check/cpp/generate_ast.cpp [hunk 9]
```diff
@@ -686,6 +708,40 @@ auto GenerateAst(Context& context,
...
+                   source_code);
+
+  auto buffer = llvm::MemoryBuffer::getMemBufferCopy(code_stream.TakeStr(),
+                                                     "<inline c++>");
+  clang::FileID file_id =
+      preprocessor.getSourceManager().createFileID(std::move(buffer));
+
+  if (preprocessor.EnterSourceFile(file_id, nullptr, clang::SourceLocation())) {
+    // Clang will have generated a suitable error. There's no anything more to
+    // do here.
```
Comment (zygoloid) [inline toolchain/check/cpp/generate_ast.cpp]:
```suggestion
    // Clang will have generated a suitable error. There's nothing more to do
    // here.
```
